Update Output Manager UTXO query to split into multiple queries

philipr-za opened the pull request:
## Description
Previously when the output manager service would send its UTXO validity query to its Base Node it would send a single query with all the UTXO hashes in it. The Base Node then responds with a single response message with all the valid UTXO’s that match a hash in the queries list. In large wallets this caused a problem where the response message could become large enough to be rejected by the frame size limits in the comms stack.

In order to mitigate this issue the Output Manager service will now send queries consisting of up to a maximum number of outputs (specified in the config) and will send multiple queries until all the outputs have been queried so that any one response does not hit the frame size limit.

## How Has This Been Tested?
`test_startup_utxo_scan` was updated to have more UTXO's than the max query limit and tested response to the 2 part query that was generated.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

sdbondi commented on base_layer/wallet/src/output_manager_service/config.rs:
```diff
@@ -25,12 +25,14 @@ use std::time::Duration;
     fn default() -> Self {
         Self {
             base_node_query_timeout: Duration::from_secs(30),
+            max_utxo_query_size: 5000,
```
Noting the max request message size would be `160kb`  and each response would be roughly less than `2Mb`

philipr-za replied:
Yes that lines up with what I observed.

sdbondi commented on base_layer/wallet/src/output_manager_service/service.rs:
```diff
@@ -450,46 +450,120 @@ where
+                // Determine how many rounds of base node request we need to query all the outputs in batches of
+                // max_utxo_query_size
+                let rounds =
+                    ((unspent_outputs.len() as f32) / (self.config.max_utxo_query_size as f32 + 0.1)) as usize + 1;
```
Maybe use `ceil()` here? Otherwise, wouldn't you get an extra round if number if `unspent_outputs.len()` is a multiple of `self.config.max_utxo_query_size`? 
```suggestion
                    ((unspent_outputs.len() as f32) / (self.config.max_utxo_query_size as f32 + 0.1)).ceil() as usize;
```

philipr-za replied:
Could actually use ceil but it will be the same, there is a sneaky +0.1 in the denominator to avoid the issue you described but Ceil is neater.

sdbondi commented on base_layer/wallet/src/output_manager_service/service.rs:
```diff
@@ -450,46 +450,120 @@ where
+
+                    // Here we are going to spawn a non-blocking task that will monitor and log the progress of the
+                    // send process.
+                    let owned_request_key = request_key;
```
`u64` is a `Copy` type so you can just use it in the async block

philipr-za replied:
You can't, it complained that it was dropped too early when I used it below.

philipr-za replied:
Huh, you are right. I must of left out the `move` when I first wrote this.

sdbondi commented on base_layer/wallet/src/output_manager_service/service.rs:
```diff
@@ -450,46 +450,120 @@ where
     pub async fn query_unspent_outputs_status(
```
Few methods are public when they don't need to be?

philipr-za replied:
you are right, in fact almost all these functions don't need to be pub. A hold over from before the service model. Will remove.

sdbondi reviewed (commented):
This should be pretty efficient because requests and responses are sent/received as fast as each side is able. Unfortunately, the code complexity to do this is rather high - something to improve on in the lower layer.

Just one comment which I think points out a bug.

philipr-za committed:
Update Output Manager UTXO query to split into multiple queries

Previously when the output manager service would send its UTXO validity query to its Base Node it would send a single query with all the UTXO hashes in it. The Base Node then responds with a single response message with all the valid UTXO’s that match a hash in the queries list. In large wallets this caused a problem where the response message could become large enough to be rejected by the frame size limits in the comms stack.

In order to mitigate this issue the Output Manager service will now send queries consisting of up to a maximum number of outputs (specified in the config) and will send multiple queries until all the outputs have been queried so that any one response does not hit the frame size limit.

sdbondi reviewed (approved):
LGTM
